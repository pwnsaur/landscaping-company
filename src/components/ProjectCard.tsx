import React, { useRef } from 'react';
import Image from 'next/image';
import Link from 'next/link';
import styled from 'styled-components';
import { TypeProject } from '@/types/contentfulTypes';
import useOnScreen from '@/utils/hooks/useOnScreen';

const ProjectCard = ({
  project,
  priority,
}: {
  project: TypeProject;
  priority: boolean;
}) => {
  const { coverImage, slug, title } = project.fields;

  const ref = useRef<HTMLDivElement>(null);
  const onScreen = useOnScreen(ref);
  const shouldLoadImage = priority || onScreen;

  return (
    <StyledCard ref={ref}>
      <Link href={`/projects/${slug}`}>
        {shouldLoadImage ? (
          <ImageContainer>
            <StyledImage
              src={`https:${coverImage.fields.file.url}`}
              alt='cover-image'
              height={coverImage.fields.file.details.image!.height / 4}
              width={coverImage.fields.file.details.image!.width / 4}
              // layout='fill'
              priority={priority}
            />
          </ImageContainer>
        ) : (
          <Placeholder
            height={coverImage.fields.file.details.image!.height / 4}
          />
        )}
        <Title>{title}</Title>
      </Link>
    </StyledCard>
  );
};

export default ProjectCard;

const Placeholder = styled.div<{ height: number }>`
  width: 100%;
  height: ${({ height }) => `${height}px`};
  background-color: #eee;
`;

const StyledCard = styled.div`
  position: relative;
  height: fit-content;
  text-align: center;
  background-color: #ededed;
  transition: transform 0.2s ease-in-out;

  &:hover {
    transform: scale(1.02);
  }
`;

const ImageContainer = styled.div`
  position: relative;
  padding-bottom: 75%;
  overflow: hidden;
`;

const StyledImage = styled(Image)`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  object-fit: cover;
`;

const Title = styled.h3`
  font-size: clamp(0.8rem, 1.1vw, 1rem);
  font-weight: 500;
  margin: 0.6rem;
  text-transform: uppercase;
`;
