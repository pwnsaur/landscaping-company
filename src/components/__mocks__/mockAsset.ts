import { Asset } from 'contentful';

export const createMockAsset = (): Asset => {
  return {
    sys: {
      id: '1',
      type: 'Asset',
      createdAt: '2023-04-27T12:00:00Z',
      updatedAt: '2023-04-27T12:00:00Z',
      locale: 'en-US',
      contentType: {
        sys: {
          id: 'asset',
          type: 'Link',
          linkType: 'ContentType',
        },
      },
    },
    fields: {
      title: 'jonathan-kemper-kYl3Yp2hwtQ-unsplash',
      description: '',
      file: {
        url: '//images.ctfassets.net/sandie-clarke-q13Zq1Jufks-unsplash.jpg',
        details: {
          size: 500939,
          image: {
            width: 1920,
            height: 1080,
          },
        },
        fileName: '',
        contentType: 'image/jpeg',
      },
    },
    metadata: {
      tags: [],
    },
    toPlainObject: () => ({} as Asset),
  };
};
