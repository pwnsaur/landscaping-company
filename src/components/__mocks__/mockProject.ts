import { createMockAsset } from '@components/__mocks__/mockAsset';
import { BLOCKS } from '@contentful/rich-text-types';
import { Entry } from 'contentful';

import { TypeProjectFields } from '@/types/contentfulTypes';

export const createMockProject = (): Entry<TypeProjectFields> => {
  return {
    sys: {
      id: '1',
      type: 'Entry',
      createdAt: '2023-04-27T12:00:00Z',
      updatedAt: '2023-04-27T12:00:00Z',
      locale: 'en-US',
      contentType: {
        sys: {
          id: 'project',
          type: 'Link',
          linkType: 'ContentType',
        },
      },
    },
    fields: {
      title: 'Fifteenth Project',
      slug: 'fifteenth-project',
      content: {
        nodeType: BLOCKS.DOCUMENT,
        data: {},
        content: [],
      },
      coverImage: createMockAsset(),
      date: '2023-04-27',
    },
    metadata: {
      tags: [],
    },
    toPlainObject: () => ({} as Entry<TypeProjectFields>),
    update: () => Promise.resolve({} as Entry<TypeProjectFields>),
  };
};
