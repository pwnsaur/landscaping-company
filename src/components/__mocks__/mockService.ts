import { createMockAsset } from '@components/__mocks__/mockAsset';
import { BLOCKS } from '@contentful/rich-text-types';
import { Entry } from 'contentful';

import { TypeServiceFields } from '@/types/contentfulTypes';

export const createMockService = (): Entry<TypeServiceFields> => {
  return {
    sys: {
      id: '1',
      type: 'Entry',
      createdAt: '2023-04-27T12:00:00Z',
      updatedAt: '2023-04-27T12:00:00Z',
      locale: 'en-US',
      contentType: {
        sys: {
          id: 'service',
          type: 'Link',
          linkType: 'ContentType',
        },
      },
    },
    fields: {
      title: 'Service',
      slug: 'service',
      excerpt: 'service',
      description: {
        nodeType: BLOCKS.PARAGRAPH,
        data: {},
        content: [],
      },
      coverImage: createMockAsset(),
    },
    metadata: {
      tags: [],
    },
    toPlainObject: () => ({} as Entry<TypeServiceFields>),
    update: () => Promise.resolve({} as Entry<TypeServiceFields>),
  };
};
