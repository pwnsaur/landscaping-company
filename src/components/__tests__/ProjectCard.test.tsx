import React from 'react';
import { render } from '@testing-library/react';
import { createMockProject } from '../__mocks__/mockProject';
import ProjectCard from '@/components/ProjectCard';
import renderer from 'react-test-renderer';

const mockProject = createMockProject();

describe('ProjectCard', () => {
  test('renders correctly', () => {
    const { getByText } = render(
      <ProjectCard project={mockProject} priority={false} />
    );

    expect(getByText(mockProject.fields.title)).toBeInTheDocument();

    const tree = renderer
      .create(<ProjectCard project={mockProject} priority={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
