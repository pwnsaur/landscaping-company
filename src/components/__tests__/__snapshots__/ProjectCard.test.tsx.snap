// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCard renders correctly 1`] = `
<div
  className="sc-17c684df-1 gtilHR"
>
  <a
    href="/projects/fifteenth-project"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="sc-17c684df-0 eLMYjF"
      height={270}
    />
    <h3
      className="sc-17c684df-4 gWtXxm"
    >
      Fifteenth Project
    </h3>
  </a>
</div>
`;
